### PR TITLE
Registries Setup

### PR DESCRIPTION
**Pull request by akshatmittal**
- [x] Version Registry
  - [x] Upgrade Changes (Main & Components)
- [x] Asset Plugin Registry
  - [x] Validations for new registrations
  - [x] Validations for existing registrations
  - [x] Optional upgrade enforcement
- [x] DAO Fee Registry
  - [x] Distributor Integration

### CODE REVIEW

File: contracts/p1/Distributor.sol [hunk 1]
```diff
@@ -201,6 +233,33 @@ contract DistributorP1 is ComponentP1, IDistributor {
         require(rTokenDist != 0 || rsrDist != 0, "no distribution defined");
     }
 
+    function _validateDistributionTable() internal view {
+        uint256 dLen = destinations.length();
+        DAOFeeRegistry daoFeeRegistry = main.daoFeeRegistry();
+        require(destinations.length() > 0, "no destinations");
```
Comment (pmckelvy1) [inline contracts/p1/Distributor.sol:239]: already have dLen

File: contracts/p1/Distributor.sol [hunk 2]
```diff
@@ -201,6 +233,33 @@ contract DistributorP1 is ComponentP1, IDistributor {
...
+        uint24 daoTotals;
+
+        for (uint256 i = 0; i < dLen; ++i) {
+            address dest = destinations.at(i);
+            RevenueShare storage share = distribution[dest];
+            revTotals.rTokenTotal += share.rTokenDist;
+            revTotals.rsrTotal += share.rsrDist;
+
+            if (dest == DAO_FEE) {
+                daoTotals += share.rsrDist;
```
Comment (pmckelvy1) [inline contracts/p1/Distributor.sol:251]: i think this is still an open question. would be better to just allow veRSR to set the DAO distribution entirely

File: contracts/p1/Distributor.sol [hunk 3]
```diff
@@ -201,6 +233,33 @@ contract DistributorP1 is ComponentP1, IDistributor {
...
+            }
+        }
+
+        require(
```
Comment (pmckelvy1) [inline contracts/p1/Distributor.sol:255]: i think there's a better way to do this.  instead of having stRSR set the distribution %'s and hoping they match what the daoFeeRegistry is expecting, we could have `_setDistribution` call the daoFeeRegistry for the % cut that it is requiring from both rsrDist and rTokenDist for this particular RToken, and then calculate what that works out to, and set the daoFee dist accordingly.

for efficiency, we can remove the singular `setDistribution` and only have 'setDistributions`, so that the final daoFee calculation can be done once everything else is set.

File: contracts/p1/Main.sol [hunk 4]
```diff
@@ -58,14 +92,74 @@ contract MainP1 is Versioned, Initializable, Auth, ComponentRegistry, UUPSUpgrad
...
+
+    function upgradeRTokenTo(bytes32 versionHash, bool validateProposal) external onlyRole(OWNER) {
+        require(address(versionRegistry) != address(0), "no registry");
+        require(keccak256(abi.encodePacked(this.version())) == versionHash, "upgrade main first");
+
+        Implementations memory implementation = versionRegistry.getImplementationForVersion(
+            versionHash
+        );
+
+        if (validateProposal) {
```
Comment (pmckelvy1) [inline contracts/p1/Main.sol:118]: can't we assume that the registry is valid before the upgrade? or is this just to cover the case where the RToken is initially upgrading to 4.0.0?

File: contracts/registry/DAOFeeRegistry.sol [hunk 5]
```diff
@@ -0,0 +1,27 @@
+// SPDX-License-Identifier: BlueOak-1.0.0
+pragma solidity 0.8.19;
+
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+
+contract DAOFeeRegistry is Ownable {
```
Comment (pmckelvy1) [inline contracts/registry/DAOFeeRegistry.sol:6]: i think there are a couple of open discussion points that we should support so as to be as generic and flexible as possible:
- fee's should be applied on a per-rtoken basis
- fee's can be applied to both rsr and rtoken revenue

File: contracts/registry/VersionRegistry.sol [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            revert VersionRegistry__InvalidRegistration();
+        }
+
+        deployments[versionHash] = deployer;
+        latestVersion = versionHash;
+
+        emit VersionRegistered(versionHash, deployer);
+    }
+
+    function deprecatedVersion(bytes32 versionHash) external onlyOwner {
```
Comment (pmckelvy1) [inline contracts/registry/VersionRegistry.sol:47]: `deprecateVersion`

**Review by pmckelvy1 — CHANGES_REQUESTED** [review]

File: contracts/p1/Distributor.sol (hunk 3 above)
Comment (akshatmittal) [reply contracts/p1/Distributor.sol:255]: We don't want to remove a function, but we'd usually use `setDistributions`. The reason we're passing the values is because (1) they can choose to pay more than required; and (2) the calculation is gas heavy to be run at the end so better to let it be computed off-chain.

File: contracts/p1/Main.sol (hunk 4 above)
Comment (akshatmittal) [reply contracts/p1/Main.sol:118]: It's for _any_ upgrade including from/to 4.0.0, hence it's optional in both directions.

File: contracts/registry/DAOFeeRegistry.sol (hunk 5 above)
Comment (akshatmittal) [reply contracts/registry/DAOFeeRegistry.sol:6]: The fee is applied to _all_ revenue but is paid in RSR, that's how it's configured right now. We don't want the veRSR DAO to be holding RTokens.

File: contracts/p1/Distributor.sol (hunk 3 above)
Comment (tbrent) [reply contracts/p1/Distributor.sol:255]: @akshatmittal what I said earlier about not wanting to change the interface was a bad reason given this is 4.0.0

Something else we didn't consider is that under the currently-implemented design, if the DAO fee is raised, there has to be a push to get RTokens to call their `setDistributions()` functions. They may not want to, though. 


Possible other design: public() function executed at end of `setDistributions()` and externally callable only by veRSR

File: contracts/p1/Main.sol [hunk 7]
```diff
@@ -49,6 +56,33 @@ contract MainP1 is Versioned, Initializable, Auth, ComponentRegistry, UUPSUpgrad
         stRSR.payoutRewards();
     }
 
+    /// Set Version Registry
+    /// @dev Can only be called once.
+    function setVerionRegistry(VersionRegistry versionRegistry_) external onlyRole(OWNER) {
```
Comment (tbrent) [inline contracts/p1/Main.sol:61]: typo -- setVersionRegistry

File: contracts/p1/Main.sol (hunk 4 above)
Comment (tbrent) [reply contracts/p1/Main.sol:118]: It does seem like we need to pull apart before and after the upgrade

File: contracts/registry/DAOFeeRegistry.sol (hunk 5 above)
Comment (tbrent) [reply contracts/registry/DAOFeeRegistry.sol:6]: I was also thinking it would be okay to only do RSR, but I can see why we'd want to have generality

File: contracts/registry/VersionRegistry.sol (hunk 6 above)
Comment (tbrent) [inline contracts/registry/VersionRegistry.sol:47]: typo -- `deprecateVersion()`

File: contracts/p1/Distributor.sol (hunk 3 above)
Comment (akshatmittal) [reply contracts/p1/Distributor.sol:255]: Let me rephrase, I agree with don't need to keep the same interface here, but I see no harm in keeping the old interface and also adding a new one.

I think I like the idea that veRSR distribution is _dynamic_ and always comes from the Fee Registry in real time, so it's never actually added to the distributions but rather something that's always present regardless of the entire table.

File: contracts/p1/Main.sol (hunk 7 above)
Comment (akshatmittal) [reply contracts/p1/Main.sol:61]: Remind me it's not a good idea to write code too late into the day.

File: contracts/p1/Distributor.sol (hunk 3 above)
Comment (tbrent) [reply contracts/p1/Distributor.sol:255]: I like this direction

Comment (pmckelvy1) [reply contracts/p1/Distributor.sol:255]: yup this makes sense to me

File: contracts/p1/Main.sol (hunk 4 above)
Comment (pmckelvy1) [reply contracts/p1/Main.sol:118]: imo:
- required after the proposal
- optional before

File: contracts/p1/AssetRegistry.sol [hunk 8]
```diff
@@ -166,6 +166,25 @@ contract AssetRegistryP1 is ComponentP1, IAssetRegistry {
...
+
+            require(
+                assetPluginRegistry.isValidAsset(
+                    keccak256(abi.encodePacked(this.version())),
+                    address(asset)
+                ),
+                "unsupported asset"
+            );
+        }
+    }
```
[the remark this answers is not in the document]
Comment (akshatmittal) [reply contracts/p1/AssetRegistry.sol:186]: I considered it, but the difference isn't much since it's basically the same loop, and this is cleaner. Lmk happy to change though.

File: contracts/registry/DAOFeeRegistry.sol [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+    function getFeeDetails(address rToken)
+        external
+        view
+        returns (address recipient, uint256 feeNumerator)
+    {
+        recipient = feeRecipient;
+        feeNumerator = rTokenFeeNumerator[rToken] == 0
+            ? defaultFeeNumerator
+            : rTokenFeeNumerator[rToken];
```
Comment (tbrent) [inline contracts/registry/DAOFeeRegistry.sol:44]: I believe this makes it impossible to have a 0% fee at the beginning

File: contracts/p1/Distributor.sol [hunk 10]
```diff
@@ -94,24 +114,23 @@ contract DistributorP1 is ComponentP1, IDistributor {
...
-            uint256 totalShares = isRSR ? revTotals.rsrTotal : revTotals.rTokenTotal;
+            totalShares = isRSR ? revTotals.rsrTotal : revTotals.rTokenTotal;
             if (totalShares != 0) tokensPerShare = amount / totalShares;
             require(tokensPerShare != 0, "nothing to distribute");
         }
-
         // Evenly distribute revenue tokens per distribution share.
         // This rounds "early", and that's deliberate!
 
         Transfer[] memory transfers = new Transfer[](destinations.length());
```
Comment (tbrent) [inline contracts/p1/Distributor.sol:129]: this needs to be `destinations.length() + 1`

File: contracts/p1/Distributor.sol [hunk 11]
```diff
@@ -94,24 +114,23 @@ contract DistributorP1 is ComponentP1, IDistributor {
...
         uint256 numTransfers;
 
-        address furnaceAddr = address(furnace); // gas-saver
-        address stRSRAddr = address(stRSR); // gas-saver
```
Comment (tbrent) [inline contracts/p1/Distributor.sol:112]: I assume this is for stack-too-deep reasons

File: contracts/p1/Distributor.sol (hunk 10 above)
Comment (tbrent) [reply contracts/p1/Distributor.sol:129]: but actually I think it might be simpler to keep it size `destinations.length()` and just move the DAO fee block after the transfers. don't actually need to save it since we know it's the last one

File: contracts/registry/DAOFeeRegistry.sol [hunk 12]
```diff
@@ -0,0 +1,46 @@
+// SPDX-License-Identifier: BlueOak-1.0.0
+pragma solidity 0.8.19;
+
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+
+contract DAOFeeRegistry is Ownable {
+    address private feeRecipient;
+    uint256 private defaultFeeNumerator; // 1e4 = 100% fee
+    mapping(address => uint256) private rTokenFeeNumerator;
```
Comment (tbrent) [inline contracts/registry/DAOFeeRegistry.sol:9]: let's create a `FEE_DENOMINATOR` constant?

File: contracts/p1/Distributor.sol [hunk 13]
```diff
@@ -160,6 +194,13 @@ contract DistributorP1 is ComponentP1, IDistributor {
             revTotals.rTokenTotal += share.rTokenDist;
             revTotals.rsrTotal += share.rsrDist;
         }
+
+        // DAO Fee
+        (, uint256 feeNumerator) = main.daoFeeRegistry().getFeeDetails(address(rToken));
+        revTotals.rsrTotal += uint24(
+            (feeNumerator * uint256(revTotals.rTokenTotal + revTotals.rsrTotal)) /
+                (1e4 - feeNumerator)
```
Comment (tbrent) [inline contracts/p1/Distributor.sol:202]: should read the 1e4 out from `daoFeeRegistry()`, or could add it as a return value to `getFeeDetails()`. but shouldn't just in-line the 1e4 imo

File: contracts/registry/DAOFeeRegistry.sol (hunk 12 above)
Comment (tbrent) [reply contracts/registry/DAOFeeRegistry.sol:9]: or return it in `getFeeDetails()`

File: contracts/registry/DAOFeeRegistry.sol (hunk 9 above)
Comment (akshatmittal) [reply contracts/registry/DAOFeeRegistry.sol:44]: Not impossible right, since defaultFee can be set to 0?

File: contracts/p1/Distributor.sol (hunk 11 above)
Comment (akshatmittal) [reply contracts/p1/Distributor.sol:112]: Yep

File: contracts/registry/DAOFeeRegistry.sol (hunk 12 above)
Comment (akshatmittal) [reply contracts/registry/DAOFeeRegistry.sol:9]: Yep, wanted to do that but I see that we use this 1e4 pattern elsewhere in the protocol so didn't do it assuming it was documented globally somewhere.

Comment (akshatmittal) [reply contracts/registry/DAOFeeRegistry.sol:9]: Nothing specifically that this does assume that the same 1e4 denominator is used everywhere for revenue purposes.

File: contracts/registry/DAOFeeRegistry.sol (hunk 9 above)
Comment (tbrent) [reply contracts/registry/DAOFeeRegistry.sol:50]: Sure -- but wouldn't be able to set individual RToken fees to 0 if the default were nonzero

Comment (tbrent) [reply contracts/registry/DAOFeeRegistry.sol:50]: Might be okay

File: contracts/registry/DAOFeeRegistry.sol (hunk 12 above)
Comment (tbrent) [reply contracts/registry/DAOFeeRegistry.sol:11]: right -- forgot we already had a constant for it in `IDistributor`: `MAX_DISTRIBUTION`

File: contracts/registry/DAOFeeRegistry.sol (hunk 9 above)
Comment (akshatmittal) [reply contracts/registry/DAOFeeRegistry.sol:50]: Fairly easy to add a "isFeeSet" mapping so lmk

**Comment by tbrent** [comment]
New protocol requirement being introduced with this PR: distributor table must sum to beyond 1e4. This is to ensure the DAO fee has enough precision.

File: contracts/p1/AssetRegistry.sol [hunk 14]
```diff
@@ -166,6 +166,25 @@ contract AssetRegistryP1 is ComponentP1, IAssetRegistry {
         }
     }
 
+    /// @inheritdoc IAssetRegistry
+    function validateCurrentAssets() external view {
```
Comment (julianmrodri) [inline contracts/p1/AssetRegistry.sol:170]: We should add a check here to revert elegantly if the registries are not set up. Otherwise it ends up reverting with a " fuinction returns unexpected amount of data" when trying to return the value for the first item, whent he Asset Plugin Registry address is 0 for example

**Review by julianmrodri — CHANGES_REQUESTED** [review]

File: test/Main.test.ts [hunk 15]
```diff
@@ -1167,6 +1167,10 @@ describe(`MainP${IMPLEMENTATION} contract`, () => {
       ).to.be.revertedWith('not enough gas to unregister safely')
     })
 
+    it('Should validate current assets if no Plugin Registry', async () => {
+      expect(await assetRegistry.validateCurrentAssets()).to.not.be.reverted
```
Comment (akshatmittal) [inline test/Main.test.ts:1171]: Shouldn't this be `await expect(assetRegistry...` or am I confusing this?
